File: tree/lca.cpp
#include <cstddef>
#include <unordered_map>
#include <stack>
#include <set>
#include <iostream>

using namespace std;

/*
Question:
Given a binary tree and two nodes in it p and q, return their lowest common ancestor
in the tree.

Solution:
Create a parent-child map as you traverse the tree. Once you find both p and q, travel
up their chain of parents until you find the first intersection.

*/

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
  TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q)
  {
    if (!root || root == p || root == q)
      return root;
    // initialize parent mapping
    unordered_map<TreeNode *, TreeNode *> parent;
    parent[root] = NULL;
    // initialize stack for dfs
    stack<TreeNode *> stk;
    stk.push(root);
    // search tree until you find p and q
    while (!parent[p] || !parent[q])
    {
      // search left and right, populating parent map as we go
      TreeNode *node = stk.top();
      stk.pop();
      if (node->left)
      {
        parent[node->left] = node;
        stk.push(node->left);
      }
      if (node->right)
      {
        parent[node->right] = node;
        stk.push(node->right);
      }
    }

    set<TreeNode *> ancestor;
    // insert all parents of p into ancestor set
    while (p)
    {
      ancestor.insert(p);
      p = parent[p];
    }
    // find the first ancestor of q which is also an ancestor of p
    while (ancestor.find(q) == ancestor.end())
    {
      q = parent[q];
    }
    return q;
  }
};
// this unit test doesn't quite work
int main()
{
  TreeNode *zero = & TreeNode(0);
  TreeNode *one = &TreeNode(1);
  TreeNode *two = &TreeNode(2);
  TreeNode *three = &TreeNode(3);
  zero->left = one;
  one->left = two;
  one->right = three;

  Solution solution;
  // expected value: 1
  cout << solution.lowestCommonAncestor(zero, two, three)->val;
}

File: graph/search/dfs/lca.java
package graph.search.dfs;

import java.util.*;

/*
Q: Given a binary tree and two nodes of the tree, return the node which is the least
common ancestor of the two.

S: DFS for p and q, while recording a map of child->parent. Once we've found both p and q,
form a set of all the ancestors of p. Then move q up the tree until it coincides with
an ancestor of p.

Script:
- Are we guaranteed that the inputs are not null?
- Let's handle the null case by throwing an exception.
- Are we guaranteed that p and q are in the tree?
- Yes


- Let me give a quick overview of the problem.
- We want to search the tree for p and q. Having found them, we want to
travel back up the parents until we find the first intersection.
- In order to travel back up the tree, we will maintain a child->parent mapping
as we do the initial search.
- Having found both nodes, we can travel up the parents to create a set of 'ancestors'
of one of the nodes.
- Then we can travel up the other node until we find an intersection with the 'ancestors' set.
- This intersection will be the LCA, which completes the solution.
- This solution will run in O(N) time, where N is the size of the tree.
- If you agree with my analysis, I can start coding it up.
*/

public class lca {
    // node class
    public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
      }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) throws IllegalArgumentException {
      if (root == null || p == null || q == null) {
        throw new IllegalArgumentException("TreeNode must not be null");
      }
      // a map of child to parent so we can traverse back up the tree
      Map<TreeNode, TreeNode> parent = new HashMap<TreeNode, TreeNode>();

      // Stack for DFS (contains the nodes we need to search, from latest to earliest)
      Stack<TreeNode> stack = new Stack<TreeNode>();
      // put a parent mapping for the root
      parent.put(root, null);
      // start the search path w/ root
      stack.push(root);

      // While we have not yet found both p and q, search for p and q
      // and populate the parent map
      while (!parent.containsKey(p) || !parent.containsKey(q)) {
        TreeNode node = stack.pop();

        // populate the parent map and search left
        if (node.left != null) {
          parent.put(node.left, node);
          stack.push(node.left);
        }

        // populate the parent map and search right
        if (node.right != null) {
          parent.put(node.right, node);
          stack.push(node.right);
        }
      }

      // set of ancestors of p. We use a set for fast lookup of duplicates
      // when we go through the ancestors of q.
      Set<TreeNode> ancestors = new HashSet<TreeNode>();
      while (p != null) {
        ancestors.add(p);
        p = parent.get(p);
      }

      // move q up the tree until we find a common ancestor with p.
      while (!ancestors.contains(q)) {
        q = parent.get(q);
      }

      // once the above while loop has ended, q is an ancestor of both p and q.
      // Hence return.
      return q;
    }
}
